### PR TITLE
add transition constraints shared across instructions to Processor Table

### LINKED ISSUE #28: Use instruction bucket's constraints in Processor Table
Even though the constraints for the various instruction groups (like `grow_stack`, `keep_stack`, etc.) are both specified and coded up, the relevant chunk of code is never actually used, i.e., the constraints are never checked. Change that.

### LINKED ISSUE #95: Drop Instruction Buckets
Instruction Buckets are cumbersome. A tiny change in the ISA throws the instruction buckets into disarray. Getting the constraints right is challenging.

I propose to get rid of instruction buckets altogether. There are common constraints among various instructions, but the code for these constraints can still be written once and reused without instruction buckets. Instead, each instruction selector selects all the constraints that apply to the given instruction – including constraints that this instruction has in common with others.

### PR DESCRIPTION
**Pull request by jan-ferdinand**
add transition constraints shared across instructions to Processor Table

- drop the concept of instruction buckets, re-use polynomials instead
- add consistency & transition constraints regarding padding
- make transition constraints respect padding rows
- hardcode opcodes, remove enum `InstructionBucket`
- drop unnecessary instruction bit

Fix bugs:
- check `osp_next` against `osp`, not against `osp_next`
- enforce update of RAM pointer for `read_mem` & `write_mem`
- instruction `halt` is `step_1`, not a jump to end of program

Improvements:
- add test printing the number of constraints for all tables

Specification:
- update specification to reflect above changes
- add opcodes, and spell out their properties
- explain mechanics of `skiz`

Fix #95

### CODE REVIEW

**Comment by jan-ferdinand** [comment]
I've delegated that part of the specification re-write effort that's not actually a _re_-write but a _write_ to issue #108.

**Review by aszepieniec — APPROVED** [review]
Looks good to me, mergit!

File: specification/src/instruction-specific-transition-constraints.md [hunk 1]
```diff
@@ -506,186 +545,80 @@ Given the high 32 bits of `st0` as `hi = st0 >> 32` and the low 32 bits of `st0`
```
[the remark this answers is not in the document]
Comment (jan-ferdinand) [reply specification/src/instruction-specific-transition-constraints.md:545]: It should, thanks.

File: triton-vm/src/table/processor_table.rs [hunk 2]
```diff
@@ -411,6 +410,46 @@ impl ExtProcessorTable {
...
+        factory: &RowPairConstraints,
+        instruction_transition_constraints: Vec<MPolynomial<XFieldElement>>,
+    ) -> Vec<MPolynomial<XFieldElement>> {
+        let ip_remains = factory.ip_next() - factory.ip();
+        let ci_remains = factory.ci_next() - factory.ci();
+        let nia_remains = factory.nia_next() - factory.nia();
+        let padding_row_transition_constraints = [
+            vec![ip_remains, ci_remains, nia_remains],
+            factory.keep_jump_stack(),
+            factory.keep_stack(),
```
[the remark this answers is not in the document]
Comment (jan-ferdinand) [reply triton-vm/src/table/processor_table.rs:424]: > I might delegate the rather lengthy refactor to future me.

See #113.

**Pull request merged**